vnext/Shared/JSI/NapiJsiRuntime.cpp

```diff
@@ -0,0 +1,1921 @@
+napi_value NapiJsiRuntime::GetHostObjectProxyHandler() {
+  if (!m_value.HostObjectProxyHandler) {
+    const napi_value handler = CreateObject();
+    SetProxyTrap<&NapiJsiRuntime::HostObjectGetTrap, 3>(handler, m_propertyId.get);
+    SetProxyTrap<&NapiJsiRuntime::HostObjectSetTrap, 4>(handler, m_propertyId.set);
+    SetProxyTrap<&NapiJsiRuntime::HostObjectOwnKeysTrap, 1>(handler, m_propertyId.ownKeys);
+    SetProxyTrap<&NapiJsiRuntime::HostObjectGetOwnPropertyDescriptorTrap, 2>(
+        handler, m_propertyId.getOwnPropertyDescriptor);
+    m_value.HostObjectProxyHandler = NapiRefHolder{this, handler};
+  }
+
+  return m_value.HostObjectProxyHandler;
+}
+
+// Sets Proxy trap method as a pointer to NapiJsiRuntime instance method.
+template <napi_value (NapiJsiRuntime::*trapMethod)(span<napi_value>), size_t argCount>
+void NapiJsiRuntime::SetProxyTrap(napi_value handler, napi_value propertyName) {
+  auto proxyTrap = [](napi_env env, napi_callback_info info) noexcept {
+    NapiJsiRuntime *runtime{};
+    napi_value args[argCount]{};
+    size_t actualArgCount{argCount};
+    CHECK_NAPI_ELSE_CRASH(
+        napi_get_cb_info(env, info, &actualArgCount, args, nullptr, reinterpret_cast<void **>(&runtime)));
+    CHECK_ELSE_CRASH(actualArgCount == argCount, "proxy trap requires argCount arguments.");
+
+    return runtime->HandleCallbackExceptions(
+        [&/*TODO: explicit captures*/]() { return (runtime->*trapMethod)(span<napi_value>(args, argCount)); });
+  };
+
+  SetProperty(handler, propertyName, CreateExternalFunction(propertyName, argCount, proxyTrap, this));
+}
+
+// The host object Proxy 'get' trap implementation.
+napi_value NapiJsiRuntime::HostObjectGetTrap(span<napi_value> args) {
+  // args[0] - the Proxy target object.
+  // args[1] - the name of the property to set.
+  // args[2] - the Proxy object (unused).
+  napi_value propertyName = args[1];
+  if (TypeOf(propertyName) == napi_symbol && StrictEquals(propertyName, m_propertyId.hostObjectSymbol)) {
+    // The special property to retrieve the target object.
+    return GetProperty(args[0], m_propertyId.hostObjectSymbol);
+  }
+  const auto &hostObject = GetJsiHostObject(args[0]);
+  PropNameIDView propertyId{this, propertyName};
+
+  return RunInMethodContext(
+      "HostObject::get", [&/*TODO: explicit captures*/]() { return GetNapiValue(hostObject->get(*this, propertyId)); });
+}
+
+// The host object Proxy 'set' trap implementation.
+napi_value NapiJsiRuntime::HostObjectSetTrap(span<napi_value> args) {
+  // args[0] - the Proxy target object.
+  // args[1] - the name of the property to set.
+  // args[2] - the new value of the property to set.
+  // args[3] - the Proxy object (unused).
+  const auto &hostObject = GetJsiHostObject(args[0]);
+  PropNameIDView propertyId{this, args[1]};
+  JsiValueView value{this, args[2]};
+  RunInMethodContext(
+      "HostObject::set", [&/*TODO: explicit captures*/]() { hostObject->set(*this, propertyId, value); });
```

Comment:
is capturing by reference, safe here? if so pls add a comment explaining why

Reply:
I'm not aware of the threading model for these.
Note, I have made captures explicit, while still keeping them by reference.

@vmoroz are these captures thread-safe?

Reply:
Yes, they are thread safe because the lamba is executed synchronously and never saved or passed to another thread.